include/seqan/blast/blast_base.h

```diff
@@ -0,0 +1,669 @@
+ * @brief translated Nucleotide Query VS Protein Subject
+ *
+ * @val BlastProgram BlastProgram::TBLASTN
+ * @brief Protein Query VS translated Nucleotide Subject
+ *
+ * @val BlastProgram BlastProgram::TBLASTX
+ * @brief translated Nucleotide Query VS translated Nucleotide Subject
+ *
+ * @val BlastProgram BlastProgram::UNKNOWN
+ * @brief Unkown type. Used internally and to signify that the type could not be inferred from the file.
+ */
+enum class BlastProgram : uint8_t
+{
+    BLASTN,         //              NUCL VS             NUCL
+    BLASTP,         //              PROT VS             PROT
+    BLASTX,         // TRANSLATED   NUCL VS             PROT
+    TBLASTN,        //              PROT VS TRANSLATED  NUCL
+    TBLASTX,        // TRANSLATED   NUCL VS TRANSLATED  NUCL
+    UNKNOWN=255
+};
+
+/*!
+ * @defgroup BlastProgramTag BlastProgram interface
+ * @brief Integral constants of @link BlastProgram @endlink
+ *
+ * @tag BlastProgramTag#BlastProgramTagBlastN
+ * @brief Tag for @link BlastProgram::BLASTN @endlink
+ * @headerfile <seqan/blast.h>
+ *
+ * @tag BlastProgramTag#BlastProgramTagBlastP
+ * @brief Tag for @link BlastProgram::BLASTP @endlink
+ * @headerfile <seqan/blast.h>
+ *
+ * @tag BlastProgramTag#BlastProgramTagBlastX
+ * @brief Tag for @link BlastProgram::BLASTX @endlink
+ * @headerfile <seqan/blast.h>
+ *
+ * @tag BlastProgramTag#BlastProgramTagTBlastN
+ * @brief Tag for @link BlastProgram::TBLASTN @endlink
+ * @headerfile <seqan/blast.h>
+ *
+ * @tag BlastProgramTag#BlastProgramTagTBlastX
+ * @brief Tag for @link BlastProgram::TBLASTX @endlink
+ * @headerfile <seqan/blast.h>
+ *
+ * @tag BlastProgramTag#BlastProgramTagUnknown
+ * @brief Tag for @link BlastProgram::UNKNOWN @endlink
+ * @headerfile <seqan/blast.h>
+ */
+
+template <BlastProgram p>
+using BlastProgramTag = std::integral_constant<BlastProgram, p>;
+
+typedef BlastProgramTag<BlastProgram::BLASTN>  BlastProgramTagBlastN;
+typedef BlastProgramTag<BlastProgram::BLASTP>  BlastProgramTagBlastP;
+typedef BlastProgramTag<BlastProgram::BLASTX>  BlastProgramTagBlastX;
+typedef BlastProgramTag<BlastProgram::TBLASTN> BlastProgramTagTBlastN;
+typedef BlastProgramTag<BlastProgram::TBLASTX> BlastProgramTagTBlastX;
+typedef BlastProgramTag<BlastProgram::UNKNOWN> BlastProgramTagUnknown;
+
+// ============================================================================
+// Functions
+// ============================================================================
+
+// ----------------------------------------------------------------------------
+// getBlastProgram()
+// ----------------------------------------------------------------------------
+
+/*
+ * Given both an enum value and a tag, return the tag's value unless it is UNKNOWN
+ * in which case the enum's value is returned. Can be used in implementations that
+ * want to cover compile-time constants and run-time settings both.
+ */
+
+// this is always evaluated at compile-time
+template <BlastProgram p>
```

Comment:
Yes, sort of. But in my case there are not ever any needless checks, if values are selected at compile-time. And even if they are dynamic, there is no recursion (which is bad because it can't be inlined and its also bad for branch prediction I think).

Seeing that there are 5 blast program formats, 3 implemented file types and legacy and plus formats for each there would already be 30 combinations of fileformats to check whenever a record a is written. This is totally not feasible.